### DISCUSSION TITLE
A entire revamped readme was launched

### DISCUSSION THREAD

**Discussion by gustavopsantos** [gustavopsantos/Reflex#28] [Announcements]
What do you guys think about new readme?
Theres is any lack? Where it can be improved?
Thanks for any contribuitions

**Comment by Lamonin** [comment]
First time understanding DI containers, so I don't know if this is normal behaviour, but the following is very confusing:
`descriptor.AddSingleton(typeof(Greeter), typeof(IStartable));`
Namely, that Greeter was treated in the container as implementing the IStartable contract, but not as a Greeter. As a result, it was not injected as Greeter and I couldn't understand why for a long time.

**Reply by gustavopsantos** [reply]
Thats a 100% expected behaviour, so, you can only resolve what you describe as contracts. So if you want to be able to resolve Greeter by its concrece implementation and also by IStartable, just do
`descriptor.AddSingleton(concrete: typeof(Greeter), contracts: typeof(Greeter), typeof(IStartable));`
But mainly you dont want to expose a concrete implementation directly, as changes will be required if you need to mock it or to create another class with same interface but different bussiness rules.

**Comment by KurbanismailovZaur** [comment]
Yes, we need to describe how exactly to create factories that will dynamically create and inject dependencies into objects. Give several different cases as an example, for example:
- A factory that can create instances of the same prefab.
- A factory that takes a prefab as a parameter and creates an instance of it.

Something like that, basically.

**Reply by gustavopsantos** [reply]
@KurbanismailovZaur Its simply a `Func<Container, T>`, you can do pretty much whatever you want. But I guess it can be confusing to some users, I will add some usage examples on readme 👍
